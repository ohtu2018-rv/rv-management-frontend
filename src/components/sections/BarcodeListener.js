import React, { Component } from 'react';
import { connect } from 'react-redux';
import { SuccessBtn } from './../buttons/Buttons';
import { setProductSelected } from '../../reducers/productReducer';
import { setBarcode, toggleBarcodeVisibility, clearBarcode } from '../../reducers/barcodeListenerReducer';
import { errorMessage } from '../../reducers/notificationReducer';
import { withRouter } from 'react-router-dom';

export class BarcodeListener extends Component {

    handleInputEvent(event) {
        this.props.setBarcode(event.target.value);
    }

    handleSubmit(event) {
        event.preventDefault();
        var product = this.props.products.find(
            prod => prod.product_barcode === this.props.barcode
        );
        if (product) {
            this.props.setProductSelected(product.product_id);
            this.props.history.push(
                `/products/${product.product_id}/stock`
            );
        } else {
            var box = this.props.boxes.find(
                b => b.box_barcode === this.props.barcode
            );
            if (box) {
                this.props.setProductSelected(box.product_id);
                this.props.history.push(
                    `/products/${box.product_id}/box`
                );
            } else {
                if (window.confirm('Not found.\nWant to create a new item?')) {
                    if (window.confirm('Create a box?')) {
                        this.props.history.push('/products/create/box');
                    } else {
                        this.props.history.push('/products/create/product');
                    }
                } else {
                    this.props.errorMessage('Product not found');
                }
            }
        }
        this.props.clearBarcode();
    }

    handleOpen(event) {
        this.props.clearBarcode();
        this.props.toggleBarcodeVisibility(true);
    }

    render() {
        return (
            <React.Fragment>
                {this.props.barcodeVisible && (
                    <form onSubmit={event => this.handleSubmit(event)}>
                        <input
                            id="barcodeInput"
                            style={{
                                paddingLeft: 16,
                                paddingRight: 16,
                                paddingTop: 11,
                                paddingBottom: 11,
                                fontSize: 16,
                                textAlign: 'center',
                                display: 'inlineBlock'
                            }}
                            ref={input => {
                                input && input.focus();
                            }}
                            value={this.props.barcode}
                            placeholder="Lue viivakoodi"
                            pattern="\d*"
                            onChange={event => this.handleInputEvent(event)}
                        />
                        <SuccessBtn
                            onClick={event => this.handleSubmit(event)}
                            type="button"
                        >
                            Hae
                        </SuccessBtn>
                    </form>
                )}
                {this.props.barcodeVisible || (
                    <SuccessBtn onClick={event => this.handleOpen(event)}>
                        Lue viivakoodi
                    </SuccessBtn>
                )}
            </React.Fragment>
        );
    }
}

const mapStateToProps = state => {
    return {
        products: state.product.products,
        boxes: state.box.boxes,
        barcode: state.barcodeListener.barcode,
        barcodeVisible: state.barcodeListener.visible
    };
};

const mapDispatchToProps = {
    setProductSelected,
    errorMessage,
    setBarcode,
    clearBarcode,
    toggleBarcodeVisibility
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(BarcodeListener));
