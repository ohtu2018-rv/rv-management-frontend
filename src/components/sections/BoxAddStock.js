import React from 'react';
import { Row, Col } from 'react-flexbox-grid';
import './styles/BoxAddStock.css';
import { connect } from 'react-redux';
import { Redirect, withRouter } from 'react-router-dom';

import { addStock } from '../../reducers/boxReducer';
import { setUpgradeStock } from '../../reducers/productReducer';

export class BoxAddStock extends React.Component {
    constructor(props) {
        super(props);

        this.calculateProductSellprice = this.calculateProductSellprice.bind(
            this
        );
        this.calculateProductSellpriceAndBoxcost = this.calculateProductSellpriceAndBoxcost.bind(
            this
        );
        this.formSubmit = this.formSubmit.bind(this);
    }

    updateFields() {
        if (this.props.box) {
            this.barcodeInput.value = this.props.box.box_barcode;
            this.productBarcodeInput.value = this.props.product.product_barcode;
            this.costInput.value = (this.props.product.buyprice / 100).toFixed(2);
            this.amountInput.value = this.props.box.items_per_box;
            this.marginInput.value = this.props.globalMargin;
            this.calculateProductSellpriceAndBoxcost();
        }
    }

    calculateProductSellprice() {
        const boxcost = parseFloat(this.boxcostInput.value);
        const amount = parseFloat(this.amountInput.value);
        const margin = parseFloat(this.marginInput.value);
        const productcost = boxcost / amount;

        this.costInput.value = productcost.toFixed(2);
        this.sellpriceInput.value = (
            productcost *
            ((100 + margin) / 100)
        ).toFixed(2);
    }

    calculateProductSellpriceAndBoxcost() {
        const productcost = parseFloat(this.costInput.value);
        const margin = parseFloat(this.marginInput.value);
        const amount = parseFloat(this.amountInput.value);

        this.boxcostInput.value = (productcost * amount).toFixed(2);
        this.sellpriceInput.value = (
            productcost *
            ((100 + margin) / 100)
        ).toFixed(2);
    }

    componentDidMount() {
        this.updateFields();
    }

    componentDidUpdate() {
        if (!this.props.upgradeStock) {
            this.updateFields();
        } else {
            const link = '/products/' + this.props.product.product_id;
            this.props.history.push(link);
        }
    }

    componentWillUnmount() {}

    formSubmit(event) {
        event.preventDefault();

        const box = {
            product_id: this.props.product.product_id,
            sellprice: this.sellpriceInput.value,
            buyprice: this.costInput.value,
            boxes: this.quantityInput.value
        };

        this.props.addStock(this.props.box.box_barcode, box, this.props.token);
    }

    render() {
        return (
            <div className="box-stocking-form">
                <form onSubmit={this.formSubmit}>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="barcode">Laatikon viivakoodi</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="barcode"
                                name="barcode"
                                type="text"
                                ref={input => {
                                    this.barcodeInput = input;
                                }}
                            />
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="cost">
                                Laatikon sisäänostohinta
                            </label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="boxcost"
                                name="boxcost"
                                type="number"
                                step="0.01"
                                min="0"
                                defaultValue="1.50"
                                ref={input => {
                                    this.boxcostInput = input;
                                }}
                                onChange={this.calculateProductSellprice}
                            />
                            <span className="unit">&euro;</span>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="margin">Kate</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="margin"
                                name="margin"
                                type="number"
                                step="1"
                                min="0"
                                defaultValue={this.props.globalMargin}
                                ref={input => {
                                    this.marginInput = input;
                                }}
                                onChange={this.calculateProductSellprice}
                            />
                            <span className="unit">%</span>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="amount">Tuotteita laatikossa</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="amount"
                                name="amount"
                                type="number"
                                step="1"
                                defaultValue="10"
                                ref={input => {
                                    this.amountInput = input;
                                }}
                                disabled
                            />
                            <span className="unit">kpl</span>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="quantity">Laatikoiden määrä</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="quantity"
                                name="quantity"
                                type="number"
                                defaultValue="1"
                                min="1"
                                step="1"
                                ref={input => {
                                    this.quantityInput = input;
                                }}
                            />
                            <span className="unit">kpl</span>
                        </Col>
                    </Row>
                    <div className="product-wrapper">
                        <Row>
                            <Col xs={3}>
                                <label className="product-title">
                                    Laatikko sisältää:
                                </label>
                            </Col>
                            <Col xs={9}>
                                <label className="product-title">
                                    {this.props.product.product_name}
                                </label>
                            </Col>
                        </Row>
                        <Row>
                            <Col xs={3}>
                                <label htmlFor="barcode">
                                    Tuotteen viivakoodi
                                </label>
                            </Col>
                            <Col xs={9}>
                                <input
                                    id="barcode"
                                    name="barcode"
                                    type="text"
                                    ref={input => {
                                        this.productBarcodeInput = input;
                                    }}
                                />
                            </Col>
                        </Row>
                        <Row>
                            <Col xs={3}>
                                <label htmlFor="cost">
                                    Tuotteen sisäänostohinta
                                </label>
                            </Col>
                            <Col xs={9}>
                                <input
                                    id="cost"
                                    name="cost"
                                    type="number"
                                    step="0.01"
                                    min="0"
                                    defaultValue="1.50"
                                    ref={input => {
                                        this.costInput = input;
                                    }}
                                    onChange={
                                        this.calculateProductSellpriceAndBoxcost
                                    }
                                />
                                <span className="unit">&euro;</span>
                            </Col>
                        </Row>
                        <Row>
                            <Col xs={3}>
                                <label htmlFor="sellprice">
                                    Tuotteen myyntihinta
                                </label>
                            </Col>
                            <Col xs={9}>
                                <input
                                    id="sellprice"
                                    name="sellprice"
                                    type="number"
                                    step="0.01"
                                    ref={input => {
                                        this.sellpriceInput = input;
                                    }}
                                />
                                <span className="unit">&euro;</span>
                            </Col>
                        </Row>
                    </div>
                    <Row>
                        <Col xs={6}>
                            <input
                                type="submit"
                                className="btn btn-success"
                                value="Osta sisään"
                            />
                        </Col>
                    </Row>
                </form>
            </div>
        );
    }
}

const mapDispatchToProps = {
    addStock,
    setUpgradeStock
};

const mapStateToProps = state => {
    return {
        globalMargin: state.product.globalMargin,
        token: state.authentication.accessToken,
        upgradeStock: state.product.upgradeStock
    };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(BoxAddStock));
