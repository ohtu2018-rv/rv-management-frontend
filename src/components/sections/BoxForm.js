import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Row, Col } from 'react-flexbox-grid';
import { Link } from 'react-router-dom';
import boxService from '../../services/boxService';
import './styles/ProductForm.css';
import {
    successMessage,
    errorMessage
} from './../../reducers/notificationReducer';
import { getProducts } from '../../reducers/productReducer';

export class BoxForm extends Component {
    formSubmit = async event => {
        event.preventDefault();
        const buyprice = this.getParsedBuyInPrice();
        const sellprice = Math.round(
            buyprice * (100 + parseInt(this.marginInput.value, 10)) / 100
        );
        const barcode = this.barcodeInput.value;
        const productCount = parseInt(this.productCountInput.value, 10);
        const boxes = parseInt(this.countInput.value, 10);
        const singleItemBuyPrice = Math.round(buyprice / productCount);
        const singleItemSellPrice = Math.round(sellprice / productCount);

        const boxProduct = this.props.products.filter(product => {
            return product.product_barcode === this.productEANInput.value;
        })[0];

        if (!boxProduct) {
            this.props.errorMessage(
                'Tuntematon tuotteen viivakoodi. Luo tuote ensin.'
            );
            return;
        }

        const parsedProduct = this.parseBoxProduct(boxProduct);

        try {
            await boxService.createBox(
                this.props.token,
                barcode,
                productCount,
                parsedProduct
            );
            await boxService.buyInBox(
                this.props.token,
                barcode,
                boxes,
                singleItemBuyPrice,
                singleItemSellPrice
            );
            this.props.getProducts(this.props.token);
            this.barcodeInput.value = '';
            this.countInput.value = 0;
            this.buyInInput.value = 0;
            this.marginInput.value = this.props.globalMargin;
            this.productEANInput.value = '';
            this.productCountInput.value = 0;
            this.props.successMessage('Lisätty uusi laatikko');
        } catch (err) {
            console.error(err);
            this.props.errorMessage('Virhe lisättäessä uutta laatikkoa');
        }
    };

    parseBoxProduct = product => {
        return {
            product_barcode: product.product_barcode,
            product_group: product.product_group,
            product_id: product.product_id,
            product_name: product.product_name,
            quantity: product.quantity,
            product_weight: 1,
            product_sellprice: product.sellprice,
            product_buyprice: product.buyprice
        };
    };

    getParsedBuyInPrice = () => {
        const buypriceArray =
            this.buyInInput.value.split('.') ||
            this.buyInInput.value.split(',');
        const buypriceString =
            buypriceArray[0] + (buypriceArray[1] + '00').substring(0, 2);
        return parseInt(buypriceString, 10);
    };

    render() {
        return (
            <div className="product-create-form">
                <form onSubmit={this.formSubmit}>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="barcode">Viivakoodi</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="barcode"
                                name="barcode"
                                type="text"
                                ref={input => {
                                    this.barcodeInput = input;
                                }}
                            />
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="count">Kappalemäärä</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="count"
                                name="count"
                                type="number"
                                defaultValue={0}
                                ref={input => {
                                    this.countInput = input;
                                }}
                            />
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="buyprice">Sisäänostohinta</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="buyprice"
                                name="buyprice"
                                type="number"
                                step="0.01"
                                min="0"
                                defaultValue={0}
                                ref={input => {
                                    this.buyInInput = input;
                                }}
                            />
                            <span className="unit">&euro;</span>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="margin">Kate</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="margin"
                                name="margin"
                                type="number"
                                step="1"
                                min="0"
                                defaultValue={this.props.globalMargin}
                                ref={input => {
                                    this.marginInput = input;
                                }}
                            />
                            <span className="unit">%</span>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="productEAN">
                                Tuotteen viivakoodi
                            </label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="productEAN"
                                name="productEAN"
                                type="text"
                                ref={input => {
                                    this.productEANInput = input;
                                }}
                            />
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="count">Kappalemäärä (tuote)</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="productcount"
                                name="productcount"
                                type="number"
                                defaultValue={0}
                                ref={input => {
                                    this.productCountInput = input;
                                }}
                            />
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={8}>
                            <input
                                type="submit"
                                className="btn btn-success"
                                value="Luo uusi laatikko"
                            />
                        </Col>
                        <Col xs={4}>
                            <Link to={'/products'}>
                                <input
                                    type="submit"
                                    className="btn btn-danger"
                                    value="Peruuta"
                                />
                            </Link>
                        </Col>
                    </Row>
                </form>
            </div>
        );
    }
}

const mapStateToProps = state => {
    return {
        globalMargin: state.product.globalMargin,
        token: state.authentication.accessToken,
        products: state.product.products
    };
};

const mapDispatchToProps = {
    successMessage,
    errorMessage,
    getProducts
};

export default connect(mapStateToProps, mapDispatchToProps)(BoxForm);
