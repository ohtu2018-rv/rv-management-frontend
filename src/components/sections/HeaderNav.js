import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import './styles/HeaderNav.css';

const activeClass = 'header-nav-active';

export class HeaderNav extends Component {
    render() {
        return (
            <div className="header-nav">
                <ul>
                    <li>
                        <NavLink to="/products" activeClassName={activeClass}>Tuotteet</NavLink>
                    </li>
                    <li>
                        <NavLink to="/users" activeClassName={activeClass}>Käyttäjät</NavLink>
                    </li>
                    <li>
                        <NavLink to="/reports" activeClassName={activeClass}>Raportit</NavLink>
                    </li>
                </ul>
            </div>
        );
    }
}

export default HeaderNav;
