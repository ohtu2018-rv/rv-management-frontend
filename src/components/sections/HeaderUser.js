import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DangerBtn } from '../buttons/Buttons';
import './styles/HeaderUser.css';

class HeaderUser extends Component {
    render() {
        return (
            <div className="header-user">
                <DangerBtn small>Kirjaudu ulos</DangerBtn>
            </div>
        );
    }
}

const mapDispatchToProps = state => {
    return {
        isAuthenticated: state.authentication.isAuthenticated
    };
};

export default HeaderUser;
