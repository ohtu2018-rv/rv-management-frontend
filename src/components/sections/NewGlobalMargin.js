import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setGlobalMargin } from '../../reducers/productReducer'

export class NewGlobalMargin extends Component {
    render() {
        return (
            <div>
                Set a new global profit margin.
                <form onSubmit={(e) => {
                    e.preventDefault()
                    this.props.setGlobalMargin(this.input.value, this.props.token)
                }}>
                    <input 
                        type="number"
                        placeholder={this.props.globalMargin}
                        ref={input => this.input = input}/>
                </form>
            </div>
        );
    }
}

const mapDispatchToProps = {
    setGlobalMargin
};

const mapStateToProps = (state) => {
    return {
        token: state.authentication.accessToken,
        globalMargin: state.product.globalMargin
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(NewGlobalMargin);
