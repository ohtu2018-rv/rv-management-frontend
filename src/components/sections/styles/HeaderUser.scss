@import "vars";

.header-user {
  margin-left: auto;
  color: #fff;
  padding: 14px 20px 13px 20px;
}
