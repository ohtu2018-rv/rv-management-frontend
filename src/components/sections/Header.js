import React, { Component } from 'react';
import { connect } from 'react-redux';
import HeaderNav from './HeaderNav';
import HeaderUser from './HeaderUser';
import { withRouter } from 'react-router-dom';

import './styles/Header.css';

export class Header extends Component {
    render() {
        return (
            <header className="topheader">
                <div className="header-title">
                    <h1>RV management</h1>
                </div>
                { this.props.isAuthenticated && <HeaderNav/> }
                { this.props.isAuthenticated && <HeaderUser/> }
            </header>
        );
    }
}

const mapStateToProps = state => {
    return {
        isAuthenticated: state.authentication.isAuthenticated
    };
};

export default withRouter(connect(mapStateToProps)(Header));
