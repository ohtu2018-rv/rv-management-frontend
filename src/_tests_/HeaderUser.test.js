import React from 'react';
import { HeaderUser } from '../components/sections/HeaderUser';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
    shallow(<HeaderUser />);
    const header = renderer.create(<HeaderUser />).toJSON();
    expect(header).toMatchSnapshot();
});

it('contains user information', () => {
    expect(shallow(<HeaderUser />).find('.header-user').length).toBe(1);
    const header = renderer.create(<HeaderUser />).toJSON();
    expect(header).toMatchSnapshot();
});
