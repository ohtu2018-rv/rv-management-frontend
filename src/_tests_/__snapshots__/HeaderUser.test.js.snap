// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contains user information 1`] = `
<div
  className="header-user"
>
  <button
    className="btn btn-danger btn-small"
    onClick={undefined}
  >
    Kirjaudu ulos
  </button>
</div>
`;

exports[`renders without crashing 1`] = `
<div
  className="header-user"
>
  <button
    className="btn btn-danger btn-small"
    onClick={undefined}
  >
    Kirjaudu ulos
  </button>
</div>
`;
